src/js/stores/document.js

```diff
@@ -166,18 +173,37 @@ define(function (require, exports, module) {
             if (suppressChange) {
                 return;
             }
-
-            // If some selected layer remains uninitialized, a new document will
-            // come along shortly. Wait until then to trigger the change event.
-            var initialized = nextDocument.layers && nextDocument.layers.selected
-                .every(function (layer) {
-                    return layer.initialized;
-                });
-
-            if (initialized) {
-                // Include the document in the event payload
-                this.emit(changeEventName || "change", nextDocument);
+            
+            var promise = Promise.bind(this);
+
+            // Notify change of layer selection state.
+            if (oldDocument) {
+                promise = promise.then(function () {
+                    oldDocument.layers.layers.forEach(function (layer) {
+                        var nextLayer = nextDocument.layers.byID(layer.id),
+                            callback = this._layerIDToStateListenner[layer.id];
```

Comment:
Yes, but "immutability" is more of a guideline than a rule :) For example, our caching scheme makes crucial use of mutating "immutable" objects. It just does so in a way that we don't expect will violate any assumption about the public interface of the immutable object.

But, again, this really might not be worth doing. Given that you only ever have one listener at a time per layer, this is so simple that whatever duplication exists isn't worth worrying about.